File: flyweight.py
class TipoLuchador:
    def __init__(self, nombre, estilo, poder):
        self.nombre = nombre
        self.estilo = estilo
        self.poder = poder
    
    def mostrar(self, vida, x, y):
        print(f"Mostrando {self.nombre} ({self.estilo}, Poder: {self.poder}) con vida {vida} en ({x}, {y})")

class FabricaLuchadores:
    _tipos = {}

    @staticmethod
    def obtener_tipo_luchador(nombre, estilo, poder):
        clave = (nombre, estilo, poder)
        if clave not in FabricaLuchadores._tipos:
            print(f"Creando nuevo tipo de luchador: {nombre} ({estilo}, Poder: {poder})")
            FabricaLuchadores._tipos[clave] = TipoLuchador(nombre, estilo, poder)
        return FabricaLuchadores._tipos[clave]

class Luchador:
    def __init__(self, tipo, vida, x, y):
        self.tipo = tipo
        self.vida = vida
        self.x = x
        self.y = y
    
    def dibujar(self):
        self.tipo.mostrar(self.vida, self.x, self.y)

# Creación de tipos de luchadores
ninja = FabricaLuchadores.obtener_tipo_luchador("Ryu", "Karate", "Hadouken")
samurai = FabricaLuchadores.obtener_tipo_luchador("Hanzo", "Kenjutsu", "Corte Relámpago")

# Creación de luchadores con diferentes posiciones y vida
luchadores = [
    Luchador(ninja, 100, 10, 20),
    Luchador(samurai, 120, 30, 40),
    Luchador(ninja, 90, 50, 60),
    Luchador(samurai, 110, 70, 80)
]

for luchador in luchadores:
    luchador.dibujar()

File: java/flyweight.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

// Clase que representa el tipo de luchador (Flyweight)
class TipoLuchador {
    private String nombre;
    private String estilo;
    private String poder;

    public TipoLuchador(String nombre, String estilo, String poder) {
        this.nombre = nombre;
        this.estilo = estilo;
        this.poder = poder;
    }

    public void mostrar(int vida, int x, int y) {
        System.out.println("Mostrando " + nombre + " (" + estilo + ", Poder: " + poder +
                           ") con vida " + vida + " en (" + x + ", " + y + ")");
    }
}

// Fabrica que gestiona los tipos de luchadores (Flyweight Factory)
class FabricaLuchadores {
    private static Map<String, TipoLuchador> tipos = new HashMap<>();

    public static TipoLuchador obtenerTipoLuchador(String nombre, String estilo, String poder) {
        String clave = nombre + "-" + estilo + "-" + poder;
        if (!tipos.containsKey(clave)) {
            System.out.println("Creando nuevo tipo de luchador: " + nombre +
                               " (" + estilo + ", Poder: " + poder + ")");
            tipos.put(clave, new TipoLuchador(nombre, estilo, poder));
        }
        return tipos.get(clave);
    }
}

// Clase Luchador que usa un TipoLuchador para dibujarse
class Luchador {
    private TipoLuchador tipo;
    private int vida;
    private int x;
    private int y;

    public Luchador(TipoLuchador tipo, int vida, int x, int y) {
        this.tipo = tipo;
        this.vida = vida;
        this.x = x;
        this.y = y;
    }

    public void dibujar() {
        tipo.mostrar(vida, x, y);
    }
}

// Clase pública con el método main (nombre del archivo: flyweight.java)
public class flyweight {
    public static void main(String[] args) {
        // Creación de tipos de luchadores (Flyweights)
        TipoLuchador ninja = FabricaLuchadores.obtenerTipoLuchador("Ryu", "Karate", "Hadouken");
        TipoLuchador samurai = FabricaLuchadores.obtenerTipoLuchador("Hanzo", "Kenjutsu", "Corte Relámpago");

        // Creación de luchadores con diferentes posiciones y vida
        List<Luchador> luchadores = new ArrayList<>();
        luchadores.add(new Luchador(ninja, 100, 10, 20));
        luchadores.add(new Luchador(samurai, 120, 30, 40));
        luchadores.add(new Luchador(ninja, 90, 50, 60));
        luchadores.add(new Luchador(samurai, 110, 70, 80));

        // Dibujar a cada luchador
        for (Luchador luchador : luchadores) {
            luchador.dibujar();
        }
    }
}
